### ISSUE TITLE
Misbehaving container that crashes the entire k8s worker node

### ISSUE THREAD

**Issue by davidbalazs** [kubernetes/kubernetes#86161]
**What happened**:
We started to see a weird behavior in Kubernetes: we have some pods that have a main container and a few sidecards, one of them being nginx. We have memory limits and requests set for all the containers (and for each container, the memory limit == memory request). 

**Sometimes, the nginx sidecar starts to spin new batches of sleeping processes (nginx workers) (up to 500 processes), causing the nginx sidecar memory usage to go up to the memory limit. When this happens, the k8s node goes into a SystemOOM state and all the pods that are running on the node become unreachable.** 

K8s event as seen on the node:
`Warning  SystemOOM  11h (x25 over 6d15h)  kubelet, ip-10-0-1-1.us-east-1.compute.internal  System OOM encountered`
The node is still reported as being ready. All the containers in the pods that are running on this node are reported as being ready as well:

```
default                poda-6fdd4c5b88-r55hp                                 5/5     Running     0          8d     
default                podb-1-0                                         6/6     Running     0          8d     
default                podc-1-2                                 5/5     Running     0          8d      
```
but the pods are unreachable, when looking at their deployments / statefulsets:
```
kubectl get deployments --all-namespaces | awk '{ if ($4 != $5) { print } }' && kubectl get sts --all-namespaces | awk '{ if (substr($3,1,1) != substr($3,3,1)) { print } }'

NAMESPACE           NAME                                                     READY   UP-TO-DATE   AVAILABLE   AGE
default                 poda     0/1     1            0           8d
NAMESPACE   NAME                                         READY   AGE
default          podb-1     4/5     8d
default     podc         2/3     8d
```

Even after I delete the deployment and the pod that has the misbehaving nginx sidecar, the other pods on the node would still be unreachable as described above.
Only after I bounce all the pods on the node, they will become reachable again.

**What you expected to happen**:
Only the pod that has the misbehaving nginx should be affected, the other pods running on the node shouldn't encounter any issue.

**How to reproduce it (as minimally and precisely as possible)**:
To quickly reproduce the issue, deploy this app to the cluster: https://github.com/davidbalazs/proc-heavy-load
This app will start lots of sleeping processes until the memory usage goes up to the limit. 

Alternatively, deploy nginx to the cluster and set the number of worker processes to be a huge one (e.g. 500)

**Anything else we need to know?**:
I have tried using --kube-reserved and --system-reserved to set up resource reservations on kubelet, but the SystemOOM still occurred. I was even trying to enforce them, using **--enforce-node-allocatable=pods,kube-reserved,system-reserved**
Relevant kubelet config that I was trying to use (note that in the podruntime slice, I had docker and kubelet running) - The values I've set are not relevant as I was trying to reserve more resources than required to see if it changes anything:
```
--kube-reserved=cpu=5,memory=10Gi,pid=10000 \
--kube-reserved-cgroup=/podruntime.slice \
--system-reserved=cpu=5,memory=10Gi,pid=10000 \
--system-reserved-cgroup=/system.slice \
--enforce-node-allocatable=pods,kube-reserved,system-reserved \
--eviction-hard=memory.available<5Gi \
```

Looking at the kubelet logs, I couldn't see anything more than the system oom event:
```
Nov 05 09:56:16 ip-10-0-1-1.us-east-1.compute.internal kubelet[6157]: I1105 09:56:16.026003    6157 manager.go:1179] Created an OOM event in container "/kubepods/burstable/pod2f60e341-618b-4d7d-b95b-6da5917ee888/f4311047445ce404d24226ddd279ad08c1aee53cf171286188cad4abbb5d8fb3"
```

As a side-note, we explicitly set the number of worker processes for nginx in its config to be 16, but the nginx spins up to 500 workers, hence the misbehavior.
**Environment**:
- Kubernetes version 
```
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-19T11:05:50Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
AWS
- OS (e.g: `cat /etc/os-release`):
CentOS 7 
- Kernel version:
 4.19.25-1.el7.centos.x86_64

**Comment by davidbalazs** [comment]
I have 2 questions about this:

1) Is there any way we can protect the stability of a worker node and the other pods on that node if one of the containers starts exhausting resources? (i.e. when it spins up lots of processes that cause the kernel on the machine to overcommit memory for example) 

2) I would also be curious to see why the other pods don't become reachable again after the systemOOM disappears (after we remove the pod that has the container that exhausts resources)

**Comment by tedyu** [comment]
For #2, could be fixed by #83455

**Comment by dims** [comment]
@davidbalazs can you use kubelet's "--pod-max-pids" that feature is designed to prevent fork bombs

**Comment by davidbalazs** [comment]
Hi @dims  
I did that as well and it works very nice for this use case, however we have some rabbitmq instances running inside our clusters. They consume a huge number of pids compared to the other apps (around 800). That seems to be because rabbbitmq uses native threads and for some reason the pids cgroup controller counts them as well when it comes to pids. If we set a max-pids limit that is higher than 800, the systemOOM would still occur
